tests/mock_module/schema.py

```diff
@@ -11,5 +11,47 @@
 
 """Example schema."""
 
+from marshmallow import Schema
+from marshmallow_utils.fields import Link
+from uritemplate import URITemplate
+
+from invenio_records_resources.resources import search_link_params, \
+    search_link_when
 # Import to make it look like we have defined our own schema.
 from invenio_records_resources.services.records.schema import RecordSchema
+
+
+class RecordLinksSchema(Schema):
+    """Schema for a record's links."""
+
+    # NOTE:
+    #   - /api prefix is needed here because above are mounted on /api
+    self = Link(
+        template=URITemplate("/api/mocks/{pid_value}"),
+        permission="read",
+        params=lambda record: {'pid_value': record.pid.pid_value}
+    )
+
+
+class SearchLinksSchema(Schema):
+    """Schema for a search result's links."""
+
+    # NOTE:
+    #   - /api prefix is needed here because api routes are mounted on /api
+    self = Link(
+        template=URITemplate("/api/mocks{?params*}"),
+        permission="search",
+        params=search_link_params(0),
+    )
+    prev = Link(
+        template=URITemplate("/api/mocks{?params*}"),
+        permission="search",
+        params=search_link_params(-1),
+        when=search_link_when(-1)
```

Comment:
It's because of this in marshmallow-utils:

```python
return factory.create_link(self.template, self.params(obj))
```

leading to 

```python
 @staticmethod
    def preprocess_vars(vars):
        """Preprocess template variables before expansion."""
        for k, v in vars.items():
            if isinstance(v, MultiDict):
                vars[k] = list(sorted(v.items(multi=True)))
        # ...
```

We can't let a `missing` be passed to `preprocess_vars` as `vars`.
Also the semantics are a little strange in `search_link_params` itself now that the schema is not a schema of parameters i.e. returning `missing` is probably what should be changed. We should probably return `{}` now. A`{'params': {}}` is not enough though to not render a link though.

Reply:
Got it. I agree the semantic is a bit strange. I would suggest to then change the functions into this, so that we don't repeat the if statement (i.e. conditions are put in the ``when``, and ``params`` only produces params):

```python
def search_link_params(page_offset):
    """Params function factory."""
    def _inner(search_dict):
        # Filter out internal parameters
        params = {
            k: v for (k, v) in search_dict.items() if not k.startswith('_')
        }
        params['page'] += page_offset
        return {'params': params}
    return _inner

def search_link_when(page_offset):
    """When function factory."""
    def _inner(search_dict):
        p = search_dict['_pagination']
        if page_offset < 0 and p.prev_page is None:
            return False
        elif page_offset > 0 and p.next_page is None:
            return False
        else:
            return True
    return _inner
```

Reply:
:+1: done